File: HW2/Problem5A/stoogesort.py
import math


def stooge_sort(array, low_index, high_index):
    n = high_index - low_index + 1

    if n == 2 and array[low_index] > array[low_index + 1]:
        storage = array[low_index]
        array[low_index] = array[low_index + 1]
        array[low_index + 1] = storage
    elif n > 2:
        m = int(math.ceil(2 * float(n) / 3))
        stooge_sort(array, low_index, m - 1 + low_index)
        stooge_sort(array, n - m + low_index, high_index)
        stooge_sort(array, low_index, m - 1 + low_index)


def main():
    data_file = open("data.txt", "r")
    out_file = open("stooge.out", "w")

    for line in data_file:
        line_array = list(map(int, line.split()))

        int_array = line_array[1:]

        stooge_sort(int_array, 0, len(int_array) - 1)

        for char in int_array:
            out_file.write(str(char) + ' ')

        out_file.write('\n')

    data_file.close()
    out_file.close()


if __name__ == "__main__":
    main()

File: HW2/Problem5A/stoogesort.js
"use strict";

const fs = require("fs");

const stoogeSort = (array, lowIndex, highIndex) => {
  const n = highIndex - lowIndex + 1;

  if (n === 2 && array[lowIndex] > array[lowIndex + 1]) {
      const storage = array[lowIndex];
      array[lowIndex] = array[lowIndex + 1];
      array[lowIndex + 1] = storage;
  }

  else if (n > 2) {
      const m = Math.floor((2 * n) / 3);
      stoogeSort(array, lowIndex, m - 1 + lowIndex);
      stoogeSort(array, n - m + lowIndex, highIndex);
      stoogeSort(array, lowIndex, m - 1 + lowIndex);
  }
};

const dataFile = fs.readFileSync("data.txt");
fs.writeFileSync("stooge.out", "");
const dataString = dataFile.toString("utf8");
const multiLineArray = dataString.split("\n");

multiLineArray.forEach(line => {
    let lineArray = line.split(" ");
    
    for(let i = 0; i < lineArray.length; i++) {
        lineArray[i] = parseInt(lineArray[i], 10);
    }

    const intArray = lineArray.slice(1);

    stoogeSort(intArray, 0, intArray.length - 1);

    intArray.forEach(number => {
      fs.appendFileSync("stooge.out", number + " ");
    });

    fs.appendFileSync("stooge.out", "\n");
});

